agent/consul/state/catalog.go

```diff
@@ -922,6 +1033,30 @@ func (s *Store) serviceNodes(ws memdb.WatchSet, serviceName string, connect bool
 		results = append(results, service.(*structs.ServiceNode))
 	}
 
+	// If we are querying for Connect nodes, the associated proxy might be a gateway.
+	// Gateways are tracked in a separate table, and we append them to the result set.
+	if connect {
```

Comment:
Ah so this is neat. You basically are hijacking the normal connect-flavored catalog reply to reply with the terminating gateways instead.

I guess you cleverly avoid an issue of recursion issue here with the TG itself needing to lookup catalog entries because by definition the TG _isn't_ going to be doing a connect catalog query because things it fronts *don't* play the Connect game already.